mgr/dashboard: do not recommend throughput for ssd's only cluster

nizamial09 opened the pull request:
This is just a bug fix where we recommend the throughput option even if
 there are only ssd's are present in the cluster.

Fixes: https://tracker.ceph.com/issues/56413
Signed-off-by: Nizamudeen A <nia@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

sunilangadi2 reviewed (approved):
Thanks @nizamial09 on the quick fix, LGTM.

epuertat commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -365,10 +367,34 @@ def reweight(self, svc_id, weight):
     def _create_predefined_drive_group(self, data):
         orch = OrchClient.instance()
         option = OsdDeploymentOptions(data[0]['option'])
+        nvme_model = None
+        ssd_model = None
+        for inventory_host in orch.inventory.list(hosts=None, refresh=True):
+            for device in inventory_host.devices.devices:
+                if device.available and device.human_readable_type == 'ssd':
+                    if '/dev/nvme' in device.path:
+                        nvme_model = device.sys_api['model']
+                    else:
+                        ssd_model = device.sys_api['model']
```
This is still a hack... What it more than 1 model is used for SSDs or NVMes? Can't we pass a path pattern to match NMVe/SSD devs?

@adk3798 do you know if we can pass globs to the path of a device?

epuertat commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -365,10 +367,34 @@ def reweight(self, svc_id, weight):
                     option]['encrypted'] = data[0]['encrypted']
+
+                if option == OsdDeploymentOptions.THROUGHPUT:
+                    if not data[0]['hdd_used']:
+                        predefined_drive_groups[
+                            option]['db_devices'].pop('rotational', 0)
+                        predefined_drive_groups[
+                            option]['data_devices'].pop('rotational', 1)
```
Maybe worthy of a comment explaining why this? I look at this code and it's hard for me to follow...

nizamial09 replied:
done!

epuertat reviewed (approved):
Quick and efficient as always, @nizamial09 ! If it works, it works... but I have a few concerns.

pereman2 commented on src/pybind/mgr/dashboard/controllers/osd.py:
```diff
@@ -509,20 +535,26 @@ def deployment_options(self):
             res.recommended_option = OsdDeploymentOptions.COST_CAPACITY
-        if ssds:
+            res.options[OsdDeploymentOptions.COST_CAPACITY].hdd_used = hdds
+        if ssds and (hdds or nvmes):
```
```suggestion
        if hdds and (ssds or nvmes)```
 I remember this being the througput case. Anyways we have to check if this is correct.

nizamial09 replied:
hmm what I remember from the throughput option is that, throughput can be `hdds/ssds` where `hdds` is `data_devices` and `ssds` are `db_devices`. it can also bee `ssds/nvmes` where `ssds` can be `data_devices` and nvmes can be `db_devices`.

nizamial09 commented:
@epuertat @pereman2 @sunilangadi2 I changed the approach to only fix the bug which is happening here and added the tests for it. please re-review this.
